Add matchmaking

smoogipoo opened the pull request:
- [x] https://github.com/ppy/osu/pull/34819

smoogipoo commented:
Pulling this out of draft as it's ready for review code-wise. I haven't done my due diligence in the OP yet, and designs are still pending.

smoogipoo committed:
Add matchmaking

peppy commented on osu.Game/Online/Multiplayer/IMultiplayerRoomServer.cs:
```diff
@@ -119,5 +119,17 @@ public interface IMultiplayerRoomServer
         /// <exception cref="UserBlocksPMsException">The invited user does not accept private messages.</exception>
         Task InvitePlayer(int userId);
+
+        /// <summary>
+        /// Raise a candidate playlist item to be played in the current round.
+        /// </summary>
+        /// <param name="playlistItemId">The playlist item.</param>
+        Task MatchmakingToggleSelection(long playlistItemId);
+
+        /// <summary>
+        /// Debug only - skips to the next stage of the matchmaking room.
+        /// </summary>
+        /// <returns></returns>
+        Task MatchmakingSkipToNextStage();
```
If there's a reason then sure, but I'd want to know the reason 😅

smoogipoo replied:
The most I'm going to split this out for now is what I've done in the recent push. That is:
- `IMultiplayerLoungeServer_Matchmaking.cs`
- `IMultiplayerRoomServer_Matchmaking.cs`

Anything further can be done post merge.

smoogipoo replied:
Scratch that - have moved things to `IMatchmakingServer`.

smoogipoo committed:
Fix event handler leak

smoogipoo committed:
Separate IMatchmakingClient and IMultiplayerClient

Co-authored-by: Dean Herbert <pe@ppy.sh>

smoogipoo committed:
Join matchmaking rooms with password

smoogipoo committed:
Add ruleset selector (#130)

smoogipoo committed:
Fix test failure

smoogipoo committed:
Query for available pools for selection (#131)

nekodex committed:
Add some audio feedback to the matchmaking flow (#132)

smoogipoo committed:
Fix results screen crash from missing users (#133)

smoogipoo committed:
Fix matchmaking being permanently sound-ducked (#134)

smoogipoo committed:
Pessimistically set the beatmap in all stages (#135)

smoogipoo committed:
Transmit availability with ready state

Regressed with https://github.com/ppy/osu-server-spectator/pull/311.

As it turns out, the method not just resets ready states but also the
beatmap availabilities. So we have to send it again here.

I have a feeling this is also broken in standard multiplayer in some way
or another.

smoogipoo committed:
Split main menu buttons into multiplayer section (#136)

Co-authored-by: Dean Herbert <pe@ppy.sh>

peppy reviewed (dismissed):
Basic pass done of surrounding code.

Won't merge until after the next release is tagged.

smoogipoo committed:
Allow exit during matchmaking intro (#137)

peppy committed:
Merge branch 'master' into matchmaking

peppy committed:
Fix main menu key tests
